engine/consensus/approvals/request_tracker.go

```diff
@@ -58,40 +61,67 @@ RequestTracker
 // Index on result ID, incorporated block ID and chunk index.
 // Is concurrency-safe.
 type RequestTracker struct {
+	headers           storage.Headers
 	index             map[flow.Identifier]map[flow.Identifier]map[uint64]RequestTrackerItem
 	blackoutPeriodMin int
 	blackoutPeriodMax int
-	lock              sync.RWMutex
+	lock              sync.Mutex
+	byHeight          map[uint64]flow.IdentifierList
+	lowestHeight      uint64
 }
 
 // NewRequestTracker instantiates a new RequestTracker with blackout periods
 // between min and max seconds.
-func NewRequestTracker(blackoutPeriodMin, blackoutPeriodMax int) *RequestTracker {
+func NewRequestTracker(headers storage.Headers, blackoutPeriodMin, blackoutPeriodMax int) *RequestTracker {
 	return &RequestTracker{
+		headers:           headers,
 		index:             make(map[flow.Identifier]map[flow.Identifier]map[uint64]RequestTrackerItem),
+		byHeight:          make(map[uint64]flow.IdentifierList),
 		blackoutPeriodMin: blackoutPeriodMin,
 		blackoutPeriodMax: blackoutPeriodMax,
 	}
 }
 
-// Get returns the tracker item for a specific chunk, and creates a new one if
-// it doesn't exist.
-func (rt *RequestTracker) Get(resultID, incorporatedBlockID flow.Identifier, chunkIndex uint64) RequestTrackerItem {
-	rt.lock.RLock()
+// TryUpdate tries to update tracker item if it's not in blackout period. Returns the tracker item for a specific chunk
+// (creates it if it doesn't exists) and whenever request item was successfully updated or not.
+// Since RequestTracker prunes items by height it can't accept items for height lower than cached lowest height.
+// If height of executed block pointed by execution result is smaller than the lowest height, sentinel mempool.DecreasingPruningHeightError is returned.
+// In case execution result points to unknown executed block exception will be returned.
+func (rt *RequestTracker) TryUpdate(result *flow.ExecutionResult, incorporatedBlockID flow.Identifier, chunkIndex uint64) (RequestTrackerItem, bool, error) {
+	resultID := result.ID()
+	rt.lock.Lock()
+	defer rt.lock.Unlock()
 	item, ok := rt.index[resultID][incorporatedBlockID][chunkIndex]
-	rt.lock.RUnlock()
 
 	if !ok {
 		item = NewRequestTrackerItem(rt.blackoutPeriodMin, rt.blackoutPeriodMax)
-		rt.Set(resultID, incorporatedBlockID, chunkIndex, item)
+		err := rt.set(resultID, result.BlockID, incorporatedBlockID, chunkIndex, item)
+		if err != nil {
+			return item, false, fmt.Errorf("could not set created tracker item: %w", err)
+		}
 	}
-	return item
+
+	canUpdate := !item.IsBlackout()
+	if canUpdate {
+		item.Update()
+		rt.index[resultID][incorporatedBlockID][chunkIndex] = item
+	}
+
+	return item, canUpdate, nil
```

Comment:
it is concurrency safe only because it's a copy. I've made this change to support concurrent usage by providing copy of object to caller

Reply:
💡 I see. That makes sense. 

What confused me was the pointer receiver on the `Update` method, which mutates the object _in place_: https://github.com/onflow/flow-go/blob/8b3376a885bffd460da93560fafe9c2388948c75/engine/consensus/approvals/request_tracker.go#L40-L48

For me it would be easier to understand the "copy-on-write pattern", if we used a value receiver for both methods and the Update would actually return a copy. Thereby the `RequestTrackerItem` itself reflects the fact that it should be immutable.